File: src/url_ingestion/fetcher.py
"""HTTP fetching with error handling."""

import requests
from requests.exceptions import ConnectionError, RequestException, Timeout

from url_ingestion.models import ErrorType, FetchError, FetchResult, FetchSuccess, NormalizedURL

USER_AGENT = "RecipeScraper/1.0"
TIMEOUT_SECONDS = 30
MAX_REDIRECTS = 5
MAX_SIZE_BYTES = 10 * 1024 * 1024  # 10MB
CHUNK_SIZE = 8192


def fetch(normalized_url: NormalizedURL) -> FetchResult:
    """Fetch HTML content from a normalized URL.

    Args:
        normalized_url: The normalized URL to fetch.

    Returns:
        FetchSuccess with HTML content, or FetchError with error details.
    """
    url = normalized_url.url

    try:
        response = requests.get(
            url,
            timeout=TIMEOUT_SECONDS,
            headers={"User-Agent": USER_AGENT},
            allow_redirects=True,
            stream=True,
        )
        response.max_redirects = MAX_REDIRECTS
    except Timeout:
        return FetchError(
            error_type=ErrorType.NETWORK,
            message="Request timed out",
            url=url,
            details={"timeout_seconds": TIMEOUT_SECONDS},
        )
    except ConnectionError as e:
        return _handle_connection_error(e, url)
    except RequestException as e:
        return FetchError(
            error_type=ErrorType.NETWORK,
            message=f"Network error: {e}",
            url=url,
            details=None,
        )

    if not 200 <= response.status_code < 300:
        return _handle_http_error(response, url)

    content_type = response.headers.get("Content-Type", "")
    if not content_type.lower().startswith("text/html"):
        return FetchError(
            error_type=ErrorType.CONTENT_TYPE,
            message=f"Expected HTML but received {content_type or 'unknown content type'}",
            url=url,
            details={"content_type": content_type},
        )

    content_result = _read_content_with_limit(response, url)
    if isinstance(content_result, FetchError):
        return content_result

    final_url = None
    if response.url != url:
        final_url = response.url

    return FetchSuccess(
        url=normalized_url,
        html=content_result,
        status_code=response.status_code,
        content_type=content_type,
        final_url=final_url,
    )


def _handle_connection_error(e: ConnectionError, url: str) -> FetchError:
    """Convert connection errors to FetchError with appropriate message."""
    error_str = str(e).lower()

    if "name or service not known" in error_str or "nodename nor servname" in error_str:
        message = "DNS resolution failed: could not resolve hostname"
    elif "connection refused" in error_str:
        message = "Connection refused: server is not accepting connections"
    elif "no route to host" in error_str:
        message = "Network unreachable: no route to host"
    else:
        message = f"Connection failed: {e}"

    return FetchError(
        error_type=ErrorType.NETWORK,
        message=message,
        url=url,
        details=None,
    )


def _handle_http_error(response: requests.Response, url: str) -> FetchError:
    """Convert HTTP error status codes to FetchError."""
    status_code = response.status_code

    if 400 <= status_code < 500:
        if status_code == 404:
            message = "Page not found (404)"
        elif status_code == 403:
            message = "Access forbidden (403)"
        elif status_code == 401:
            message = "Authentication required (401)"
        else:
            message = f"Client error: HTTP {status_code}"
    else:
        if status_code == 500:
            message = "Internal server error (500)"
        elif status_code == 502:
            message = "Bad gateway (502)"
        elif status_code == 503:
            message = "Service unavailable (503)"
        else:
            message = f"Server error: HTTP {status_code}"

    return FetchError(
        error_type=ErrorType.HTTP,
        message=message,
        url=url,
        details={"status_code": status_code},
    )


def _read_content_with_limit(response: requests.Response, url: str) -> str | FetchError:
    """Read response content with size limit.

    Args:
        response: The HTTP response to read.
        url: The URL for error reporting.

    Returns:
        The content as a string, or FetchError if size limit exceeded.
    """
    content_bytes = b""

    for chunk in response.iter_content(chunk_size=CHUNK_SIZE):
        content_bytes += chunk
        if len(content_bytes) > MAX_SIZE_BYTES:
            response.close()
            return FetchError(
                error_type=ErrorType.SIZE,
                message=f"Response exceeds {MAX_SIZE_BYTES // (1024 * 1024)}MB limit",
                url=url,
                details={"max_size_bytes": MAX_SIZE_BYTES},
            )

    encoding = response.encoding or "utf-8"
    try:
        return content_bytes.decode(encoding, errors="replace")
    except (LookupError, UnicodeDecodeError):
        return content_bytes.decode("utf-8", errors="replace")

File: src-tauri/src/url_ingestion/fetcher.rs
use reqwest::Client;
use std::time::Duration;

use crate::url_ingestion::models::{FetchError, FetchSuccess, NormalizedUrl};

/// HTTP client configuration constants
const TIMEOUT_SECONDS: u64 = 30;
const MAX_REDIRECTS: usize = 5;
const MAX_SIZE_BYTES: usize = 10_485_760; // 10MB
const USER_AGENT: &str = "RecipeScraper/1.0";

/// Creates a configured HTTP client for URL fetching.
pub fn create_client() -> Result<Client, FetchError> {
    Client::builder()
        .redirect(reqwest::redirect::Policy::limited(MAX_REDIRECTS))
        .timeout(Duration::from_secs(TIMEOUT_SECONDS))
        .user_agent(USER_AGENT)
        .build()
        .map_err(|e| FetchError::Network {
            message: format!("Failed to create HTTP client: {}", e),
            url: String::new(),
            details: Some(e.to_string()),
        })
}

/// Fetches HTML content from a URL.
///
/// Returns FetchSuccess with HTML content or FetchError on failure.
pub async fn fetch(
    client: &Client,
    normalized_url: &NormalizedUrl,
) -> Result<FetchSuccess, FetchError> {
    let url_string = normalized_url.to_url_string();

    // Make the GET request
    let response = client
        .get(&url_string)
        .send()
        .await
        .map_err(|e| map_reqwest_error(e, &url_string))?;

    // Check HTTP status
    let status = response.status();
    if status.is_client_error() {
        return Err(FetchError::Http {
            message: format_http_error(status.as_u16()),
            url: url_string,
            status_code: status.as_u16(),
        });
    }
    if status.is_server_error() {
        return Err(FetchError::Http {
            message: format_http_error(status.as_u16()),
            url: url_string,
            status_code: status.as_u16(),
        });
    }

    // Check Content-Type
    let content_type = response
        .headers()
        .get(reqwest::header::CONTENT_TYPE)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string();

    if !content_type.starts_with("text/html") {
        return Err(FetchError::ContentType {
            message: format!("Expected HTML but received {}", content_type),
            url: url_string,
            content_type,
        });
    }

    // Get final URL (after redirects)
    let final_url = response.url().to_string();
    let final_url = if final_url != url_string {
        Some(final_url)
    } else {
        None
    };

    // Read body with size limit
    let html = read_body_with_limit(response, &url_string).await?;

    Ok(FetchSuccess {
        url: normalized_url.clone(),
        html,
        status_code: status.as_u16(),
        content_type,
        final_url,
    })
}

/// Reads the response body with a size limit.
async fn read_body_with_limit(
    response: reqwest::Response,
    url: &str,
) -> Result<String, FetchError> {
    // Check Content-Length header if available
    if let Some(content_length) = response.content_length() {
        if content_length as usize > MAX_SIZE_BYTES {
            return Err(FetchError::Size {
                message: format!(
                    "Response too large: {} bytes (max {} bytes)",
                    content_length, MAX_SIZE_BYTES
                ),
                url: url.to_string(),
                max_bytes: MAX_SIZE_BYTES,
            });
        }
    }

    // Stream body and enforce size limit
    let bytes = response.bytes().await.map_err(|e| FetchError::Network {
        message: format!("Failed to read response body: {}", e),
        url: url.to_string(),
        details: Some(e.to_string()),
    })?;

    if bytes.len() > MAX_SIZE_BYTES {
        return Err(FetchError::Size {
            message: format!(
                "Response too large: {} bytes (max {} bytes)",
                bytes.len(),
                MAX_SIZE_BYTES
            ),
            url: url.to_string(),
            max_bytes: MAX_SIZE_BYTES,
        });
    }

    // Convert to UTF-8 string
    String::from_utf8(bytes.to_vec()).map_err(|e| FetchError::ContentType {
        message: format!("Response is not valid UTF-8: {}", e),
        url: url.to_string(),
        content_type: "text/html".to_string(),
    })
}

/// Maps reqwest errors to FetchError.
fn map_reqwest_error(err: reqwest::Error, url: &str) -> FetchError {
    if err.is_timeout() {
        return FetchError::Network {
            message: "Request timed out after 30 seconds".to_string(),
            url: url.to_string(),
            details: None,
        };
    }

    if err.is_connect() {
        return FetchError::Network {
            message: "Connection failed".to_string(),
            url: url.to_string(),
            details: Some(err.to_string()),
        };
    }

    // Check for DNS resolution errors
    let err_string = err.to_string().to_lowercase();
    if err_string.contains("dns")
        || err_string.contains("resolve")
        || err_string.contains("no such host")
    {
        return FetchError::Network {
            message: "DNS resolution failed".to_string(),
            url: url.to_string(),
            details: Some(err.to_string()),
        };
    }

    FetchError::Network {
        message: format!("Network error: {}", err),
        url: url.to_string(),
        details: Some(err.to_string()),
    }
}

/// Formats HTTP status code into a human-readable message.
fn format_http_error(status_code: u16) -> String {
    match status_code {
        400 => "Bad request (400)".to_string(),
        401 => "Unauthorized (401)".to_string(),
        403 => "Forbidden (403)".to_string(),
        404 => "Page not found (404)".to_string(),
        405 => "Method not allowed (405)".to_string(),
        408 => "Request timeout (408)".to_string(),
        410 => "Gone (410)".to_string(),
        429 => "Too many requests (429)".to_string(),
        500 => "Internal server error (500)".to_string(),
        502 => "Bad gateway (502)".to_string(),
        503 => "Service unavailable (503)".to_string(),
        504 => "Gateway timeout (504)".to_string(),
        _ => format!("HTTP error ({})", status_code),
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_format_http_error_404() {
        assert_eq!(format_http_error(404), "Page not found (404)");
    }

    #[test]
    fn test_format_http_error_500() {
        assert_eq!(format_http_error(500), "Internal server error (500)");
    }

    #[test]
    fn test_format_http_error_unknown() {
        assert_eq!(format_http_error(418), "HTTP error (418)");
    }
}
